lib/core/sdam/topology.js

```diff
@@ -942,11 +937,45 @@ function createAndConnectServer(topology, serverDescription) {
   server.once('connect', serverConnectEventHandler(server, topology));
   server.on('descriptionReceived', topology.serverUpdateHandler.bind(topology));
   server.on('error', serverErrorEventHandler(server, topology));
-  server.on('close', () => topology.emit('close', server));
+
+  if (connectDelay) {
+    const connectTimer = setTimeout(() => {
+      removeTimerFrom(connectTimer, topology.s.connectionTimers);
+      server.connect();
+    }, connectDelay);
+
+    topology.s.connectionTimers.push(connectTimer);
+    return server;
+  }
+
   server.connect();
   return server;
 }
 
+function resetServer(topology, serverDescription) {
+  if (!topology.s.servers.has(serverDescription.address)) {
+    return;
+  }
+
+  // first remove the old server
+  const server = topology.s.servers.get(serverDescription.address);
+  destroyServer(server, topology);
+
+  // add the new server, and attempt connection after a delay
+  const newServer = createAndConnectServer(
+    topology,
+    serverDescription,
+    topology.s.minHeartbeatFrequencyMS
+  );
+
+  topology.s.servers.set(serverDescription.address, newServer);
+}
+
+function removeTimerFrom(timer, timers) {
```

Comment:
I am somewhat inclined to make a wrapper class for managing these timers:

```js
class TimerManager {
  constructor() {
    this.timers = new Set();
  }
  setTimeout() {
    const timer = setTimeout.apply(null, arguments);
    this.timers.add(timer);
    return timer;
  }
  removeTimer(timer) {
    if (this.timers.has(timer)) {
      clearTimeout(timer);
      this.timers.remove(timer);
    }
  }
  clear() {
    this.timers.forEach(clearTimeout);
    this.timers.clear();
  }
}
```

Might be overkill, but managing a bunch of raw arrays feels a little weird. I'd at least want to switch to using `Set`s for the timers.

Reply:
done